[DM-27537] First draft proposal for Butler access control

rra opened the pull request:
Write-up of the results of our meeting plus some additional thoughts today.

rra committed:
First draft

rra committed:
Flesh out proposal, add diagram

gpdf commented:
In "Current State" I would mention that access to the database, in either case (local SQLite or database server), is mediated by SQLAlchemy.

gpdf commented on index.rst:
```diff
@@ -1,61 +1,114 @@
 
-  Subsections
-  -----------
+When a Rubin Science Platform user wants to process some data, they would call Butler to download the data, perform whatever processing they wish, and then call Butler again to store the results somewhere for future use.
```
I wouldn't say "download".  They are just trying to "access" the data in a Python environment.  If the files are in a Posix data store there really is no "downloading" involved; if they are on cloud storage in the IDF, even then I think "download" is suggesting something remote, and that's not currently a requirement.  (It could be a future one, but that's not the present situation.)

gpdf commented on index.rst:
```diff
@@ -1,61 +1,114 @@
-     Caption text.
+#. Users must be able to query the combination of Data Release and User Generated data products using Butler.
+#. Users must be able to download data products to which they have access, including the Data Release, and upload their own User Generated data products.
+#. Users must not have access to User Generated data products or the registry entries for those products unless they have been granted access by the owners of those User Generated data products.
```
Append: ", either directly or via membership in a group that has been granted access." to tie this together with the A&A requirements.

gpdf replied:
Append: ", either directly or via membership in a group that has been granted access." to tie this together with the A&A requirements.

rra committed:
Wording improvements from review

- Clarify that access can be granted via group membership
- List the object stores that support signed URLs
- Clarify that data access is not necessarily via download
- Note that Butler uses SQLAlchemy

rra commented:
Thanks for the review! I've incorporated those changes.
